src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs

```diff
@@ -988,35 +986,56 @@ public string Replace(char oldChar, char newChar)
             }
 
             // Copy the remaining characters, doing the replacement as we go.
-            ref ushort pSrc = ref Unsafe.Add(ref Unsafe.As<char, ushort>(ref _firstChar), copyLength);
-            ref ushort pDst = ref Unsafe.Add(ref Unsafe.As<char, ushort>(ref result._firstChar), copyLength);
+            ref ushort pSrc = ref Unsafe.Add(ref Unsafe.As<char, ushort>(ref _firstChar), (nint)(uint)copyLength);
+            ref ushort pDst = ref Unsafe.Add(ref Unsafe.As<char, ushort>(ref result._firstChar), (nint)(uint)copyLength);
+            nint i = 0;
 
             if (Vector.IsHardwareAccelerated && remainingLength >= Vector<ushort>.Count)
             {
-                Vector<ushort> oldChars = new Vector<ushort>(oldChar);
-                Vector<ushort> newChars = new Vector<ushort>(newChar);
+                Vector<ushort> oldChars = new(oldChar);
+                Vector<ushort> newChars = new(newChar);
 
-                do
+                Vector<ushort> original;
+                Vector<ushort> equals;
+                Vector<ushort> results;
+
+                nint lengthToExamine = (nint)(uint)(remainingLength - Vector<ushort>.Count);
+
+                if (lengthToExamine > 0)
                 {
-                    Vector<ushort> original = Unsafe.ReadUnaligned<Vector<ushort>>(ref Unsafe.As<ushort, byte>(ref pSrc));
-                    Vector<ushort> equals = Vector.Equals(original, oldChars);
-                    Vector<ushort> results = Vector.ConditionalSelect(equals, newChars, original);
-                    Unsafe.WriteUnaligned(ref Unsafe.As<ushort, byte>(ref pDst), results);
-
-                    pSrc = ref Unsafe.Add(ref pSrc, Vector<ushort>.Count);
-                    pDst = ref Unsafe.Add(ref pDst, Vector<ushort>.Count);
-                    remainingLength -= Vector<ushort>.Count;
+                    do
+                    {
+                        original = Unsafe.ReadUnaligned<Vector<ushort>>(ref Unsafe.As<ushort, byte>(ref Unsafe.Add(ref pSrc, i)));
```

Comment:
`Vector.LoadUnsafe(ref pSrc, i)` would go a long way towards readability here.

We can look at making that function public in .NET 8 to provide parity with the Vector64/128/256 functions.

Reply:
Yep, makes sense. Should this happen in this PR or separate (I image there are more places that could profit from that new API).

Reply:
I think adding it here is fine, we can then update other places to use it separately.

We are actually using "local helpers" in a few of the other places already so the readability isn't bad there today.

Reply:
Only when creating the `pSrc` and `pDst` locals, after we have a `ref ushort pSrc` everything else becomes much cleaner.

We could also provide a specialized `ref char -> ref ushort` internal overload for convenience of our own handling.

Reply:
> We could also provide a specialized ref char -> ref ushort internal overload for convenience of our own handling.

👍 at least internal would be nice to have

Reply:
Done with ed83650b7d20e41be4c3cc35ad51684153cb83dc and 5d92816266eae2c048bf95c45da7707e29acae3d